### PR TITLE
ARROW-7904: [C++][Python] Revamp metadata display, change show_metadata to verbose_metadata

### PR DESCRIPTION
**Pull request by wesm**
This is another attempt to present the information without overwhelming in the case where there is a large binary metadata blob. So the default will show just metadata keys like so:

```
foo: int32 not null, metadata.keys: ['key1']
bar: string, metadata.keys: ['key3']
-- schema.metadata.keys: ['key2']
```

Another option is to show values truncated to 50 chars (or 80 chars less the length of the key and other whitespace chars)

### CODE REVIEW

**Review by jorisvandenbossche — COMMENTED** [review]
+1 from me

As stated before, I think a truncated version would be nice (at least for the schema metadata, maybe not for the field metadata ..), but I certainly like this as well.

**Review by kou — APPROVED** [review]
+1

I also like showing truncated value too rather than showing only keys.

**Comment by bkietz** [comment]
@wesm I'm rebasing this PR to pick up the fix for https://issues.apache.org/jira/browse/ARROW-8106

File: python/pyarrow/tests/test_schema.py
Comment (bkietz) [inline python/pyarrow/tests/test_schema.py]:
```suggestion
bar: string
```

**Review by bkietz — CHANGES_REQUESTED** [review]

**Comment by wesm** [comment]
@jorisvandenbossche @kou makes sense. I'll make another iteration on this to print the truncated values so that both key and value (and any indentation) fit into 80 characters

**Comment by wesm** [comment]
Thanks @bkietz for picking this up! 

There's a small edge case when there are 0 additional characters:

```
In [10]: pa.schema([('f0', 'int32')], metadata={'key': 'value' + 'x' * 65})                                                   
Out[10]: 
f0: int32
-- metadata --
key: 'valuexxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx' + 0
```

I'll fix that and rebase. Any other feedback on this, @kou or @jorisvandenbossche want to look again?

**Comment by wesm** [comment]
I worked on this a little bit to make the Field metadata display consistently also, take a look

```
In [1]: sch = pa.schema([pa.field('f0', 'int32', metadata={'k1': 'v1'*100})],  
                        metadata={'key': 'value' + 'x' * 1000})         

In [2]: sch                                                                                                                   
Out[2]: 
f0: int32
  -- field metadata --
  k1: 'v1v1v1v1v1v1v1v1v1v1v1v1v1v1v1v1v1v1v1v1v1v1v1v1v1v1v1v1v1v1v1v1v1v1' + 132
-- schema metadata --
key: 'valuexxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx' + 935
```

This patch is good by me, I'll wait until tomorrow sometime before merging since this isn't the first patch altering the metadata display so avoiding churn would be good

**Comment by jorisvandenbossche** [comment]
One reason to maybe not show the field metadata in the table / schema repr *by default*, is that nowadays any table that was read from parquet will include field metadata ..

```
In [5]: table = pa.table(pd.DataFrame(np.random.randn(10,5), columns=list('abcde')))

In [6]: pq.write_table(table, 'test.parquet') 

In [7]: pq.read_table('test.parquet')  
Out[7]: 
pyarrow.Table
a: double
  -- field metadata --
  PARQUET:field_id: '1'
b: double
  -- field metadata --
  PARQUET:field_id: '2'
c: double
  -- field metadata --
  PARQUET:field_id: '3'
d: double
  -- field metadata --
  PARQUET:field_id: '4'
e: double
  -- field metadata --
  PARQUET:field_id: '5'
-- schema metadata --
pandas: '{"index_columns": [{"kind": "range", "name": null, "start": 0, "stop":' + 808
```

**Comment by jorisvandenbossche** [comment]
Another observation, but somewhat unrelated because I assume it takes different code paths, the Field repr in Python actually is less informative, as it doesn't show the metadata (with the example from above):

```
In [9]: pq.read_table('test.parquet').schema.field('a') 
Out[9]: pyarrow.Field<a: double>
```

(and if there is one place where I would certainly show it by default, it is here)

Regarding my comment above, it might all get complicated, if we want to have fine control over whether to show no, truncated, or full metadata, potentially separately for fields as for schema metadata. So take my comment with a grain of salt. 
But it could be an option to eg for the table repr show less by default, and for the schema repr more.

**Comment by wesm** [comment]
Okay, I'll make another iteration to add a third mode of hiding schema or field metadata and thread that through the Python reprs

**Comment by wesm** [comment]
Alright, I added options to suppress Schema and Field metadata when pretty-printing a Schema, and exposed this in Python. 

Now we have

```
In [2]: In [5]: table = pa.table(pd.DataFrame(np.random.randn(10,5), columns=list('abcde'))) 
   ...:  
   ...: In [6]: pq.write_table(table, 'test.parquet')  
   ...:  
   ...: table = pq.read_table('test.parquet')                                                                                 

In [3]: table                                                                                                                 
Out[3]: 
pyarrow.Table
a: double
b: double
c: double
d: double
e: double

In [5]: print(table.to_string(show_metadata=True))                                                                            
pyarrow.Table
a: double
  -- field metadata --
  PARQUET:field_id: '1'
b: double
  -- field metadata --
  PARQUET:field_id: '2'
c: double
  -- field metadata --
  PARQUET:field_id: '3'
d: double
  -- field metadata --
  PARQUET:field_id: '4'
e: double
  -- field metadata --
  PARQUET:field_id: '5'
-- schema metadata --
pandas: '{"index_columns": [{"kind": "range", "name": null, "start": 0, "' + 804

In [6]: table.schema                                                                                                          
Out[6]: 
a: double
  -- field metadata --
  PARQUET:field_id: '1'
b: double
  -- field metadata --
  PARQUET:field_id: '2'
c: double
  -- field metadata --
  PARQUET:field_id: '3'
d: double
  -- field metadata --
  PARQUET:field_id: '4'
e: double
  -- field metadata --
  PARQUET:field_id: '5'
-- schema metadata --
pandas: '{"index_columns": [{"kind": "range", "name": null, "start": 0, "' + 804
```

I'm going to go ahead and merge this once the builds are green and further refinements / tweaks are welcome in follow up patches

**Comment by wesm** [comment]
Merging. CI failures:

* the Python test failed due to unrelated Minio errors
* googletest failed to download (again??) in the Appveyor toolchain build

**Pull request closed without merging**
